monitor/services/src/main/java/org/datacleaner/monitor/server/controllers/ReferenceDataController.java

```diff
@@ -19,59 +19,361 @@
+
+        if (stringPattern == null) {
+            logger.warn("Could not find string pattern \"" + name + "\"");
+            throw new NoSuchResourceException();
+        }
+
+        if (request.getMethod().equals(RequestMethod.HEAD.name())) {
+            return null;
+        }
+
+        if (stringPattern instanceof SimpleStringPattern) {
+            return new StringPatternModel(name, ((SimpleStringPattern) stringPattern).getExpression(),
+                    StringPatternModel.PatternType.STRING);
+        } else if (stringPattern instanceof RegexStringPattern) {
+            final RegexStringPattern regexStringPattern = (RegexStringPattern) stringPattern;
+            return new StringPatternModel(name, regexStringPattern.getExpression(),
+                    regexStringPattern.isMatchEntireString() ? StringPatternModel.PatternType.REGEX_MATCH_ENTIRE_STRING
+                            : StringPatternModel.PatternType.REGEX);
+        } else { // RegexSwap. Treat as regex.
+            final RegexSwapStringPattern regexStringPattern = (RegexSwapStringPattern) stringPattern;
+            return new StringPatternModel(name, regexStringPattern.getRegex().getExpression(),
+                    StringPatternModel.PatternType.REGEX);
+        }
+    }
+
+    @RolesAllowed(SecurityRoles.CONFIGURATION_EDITOR)
+    @RequestMapping(value = "/stringPattern/{name}", method = { RequestMethod.PUT },
+            consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity putStringPattern(@PathVariable("tenant") final String tenant,
+            @PathVariable("name") final String name, @RequestBody final StringPatternModel stringPatternModel) {
+
+        final StringPattern stringPattern;
+        switch (stringPatternModel.getPatternType()) {
+        case STRING:
+            stringPattern = new SimpleStringPattern(name, stringPatternModel.getPattern());
+            break;
+        case REGEX:
+            stringPattern = new RegexStringPattern(name, stringPatternModel.getPattern(), false);
+            break;
+        case REGEX_MATCH_ENTIRE_STRING:
+            stringPattern = new RegexStringPattern(name, stringPatternModel.getPattern(), true);
+            break;
+        default:
+            throw new IllegalArgumentException("No such string pattern type");
+        }
+
+        final TenantContext context = _contextFactory.getContext(tenant);
+        _referenceDataDao.removeStringPattern(context, name);
+        _referenceDataDao.addStringPattern(context, stringPattern);
+
+        return ResponseEntity.created(ServletUriComponentsBuilder.fromCurrentRequestUri().build().toUri()).build();
+    }
+
+    @RolesAllowed(SecurityRoles.CONFIGURATION_EDITOR)
+    @RequestMapping(value = "/stringPattern/{name}", method = { RequestMethod.DELETE })
+    public ResponseEntity deleteStringPattern(@PathVariable("tenant") final String tenant,
```

Comment:
It's never called compile-time, so the generics doesn't really apply. Of course it still serves some documentation purpose, so I've added it (it's always `ResponseEntity<Void>` in our case).